File: tests/xcalablemp/global-view/async/C/gmove_async.c
#include <stdio.h>
#include <stdlib.h>
#include "xmp.h"

#pragma xmp nodes p(4)
#pragma xmp template t(0:7)
#pragma xmp distribute t(block) onto p

  int a[8][8];
#pragma xmp align a[i][*] with t(i)
  int b[8][8];
#pragma xmp align b[*][j] with t(j)

int main(void){

#pragma xmp loop (j) on t(j)
  for (int j = 0; j < 8; j++){
    for (int i = 0; i < 8; i++){
      b[i][j] = i * 10 + j;
    }
  }

#pragma xmp gmove async(10)
  a[:][:] = b[:][:];

#pragma xmp wait_async(10)

  int ierr = 0;

#pragma xmp loop (i) on t(i) reduction(+:ierr)
  for (int i = 0; i < 8; i++){
    for (int j = 0; j < 8; j++){
      if (a[i][j] != i * 10 + j) ierr++;
    }
  }

#pragma xmp task on p(1)
  {
    if (ierr == 0){
      printf("PASS\n");
    }
    else {
      fprintf(stderr, "ERROR\n");
      exit(1);
    }
  }

  return 0;

}

File: tests/XMP/global-view/async/F/gmove_async.f90
  program gmove_async

!$xmp nodes p(4)
!$xmp template t(8)
!$xmp distribute t(block) onto p

  integer a(8,8)
!$xmp align a(*,j) with t(j)

  integer b(8,8)
!$xmp align b(i,*) with t(i)

  integer :: ierr = 0

!$xmp loop (i) on t(i)
  do i = 1, 18
     do j = 1, 8
        b(i,j) = i * 10 + j
     end do
  end do

!$xmp gmove async(10)
  a(:,:) = b(:,:)

!$xmp wait_async(10)

!$xmp loop (j) on t(j) reduction(+:ierr)
  do j = 1, 8
     do i = 1, 8
        if (a(i,j) /= i * 10 + j) ierr = 1
     end do
  end do

!$xmp task on p(1)
  if (ierr == 0) then
     write(*,*) "PASS"
  else
     write(*,*) "ERROR"
     call exit(1)
  end if
!$xmp end task

end
